Detectors/TPC/simulation/include/TPCSimulation/SpaceCharge.h

```diff
@@ -0,0 +1,181 @@
+  void calculateLookupTables();
+  /// Update distortion and correction lookup tables by current space-charge density
+  /// \param eventTime time of current event
+  void updateLookupTables(float eventTime);
+
+  /// Set omega*tau and T1, T2 tensor terms in Langevin-equation solution
+  /// \param omegaTau omega*tau
+  /// \param t1 T1 tensor term
+  /// \param t2 T2 tensor term
+  void setOmegaTauT1T2(float omegaTau, float t1, float t2);
+  /// Set an initial space-charge density
+  /// \param hisSCDensity 3D space-charge density histogram, expected format (phi,r,z)
+  void setInitialSpaceChargeDensity(TH3* hisSCDensity);
+  /// Add ions to space-charge density
+  /// \param zPos z position
+  /// \param phiPos phi position
+  /// \param rPos radial position
+  /// \param nIons number of ions
+  void fillSCDensity(float zPos, float phiPos, float rPos, int nIons);
+  /// Propagate space-charge density along electric field by one time slice
+  void propagateSpaceCharge();
+  /// Drift ion along electric field by one time slice
+  /// \param point 3D coordinates of the ion
+  /// \return GlobalPosition3D with coordinates of drifted ion
+  GlobalPosition3D driftIon(GlobalPosition3D& point);
+
+  /// Correct electron position using correction lookup tables
+  /// \param point 3D coordinates of the electron
+  void correctElectron(GlobalPosition3D& point);
+  /// Distort electron position using distortion lookup tables
+  /// \param point 3D coordinates of the electron
+  void distortElectron(GlobalPosition3D& point);
+
+  /// Set the space-charge distortions model
+  /// \param distortionType distortion type (constant or realistic)
+  void setSCDistortionType(SCDistortionType distortionType) { mSCDistortionType = distortionType; }
+  /// Get the space-charge distortions model
+  SCDistortionType getSCDistortionType() const { return mSCDistortionType; }
+
+ private:
+  /// Convert amount of ions into charge density C/m^3
+  /// \param nIons number of ions
+  /// \return space-charge density (C/m^3)
+  float ions2Charge(int nIons);
+
+  static constexpr float DvDEoverv0 = 0.0025; //! v'(E) / v0 = K / (K*E0) for ions, used in dz calculation
+  static const float sEzField;                //! nominal drift field
+
+  static constexpr int MaxZSlices = 200;     //! default number of z slices (1 ms slices)
+  static constexpr int MaxPhiBins = 360;     //! default number of phi bins
+  static constexpr float DriftLength = 250.; //! drift length of the TPC in (cm)
+  // ion mobility K = 3.0769231 cm^2/(Vs) in Ne-CO2 90-10 published by A. Deisting
+  // v_drift = K * E = 3.0769231 cm^2/(Vs) * 400 V/cm = 1230.7692 cm/s
+  // t_drift = 250 cm / v_drift = 203 ms
+  static constexpr float IonDriftTime = 2.03e5; //! drift time of ions for one full drift (us)
+  static constexpr float RadiusInner = 85.;     //! inner radius of the TPC active area
+  static constexpr float RadiusOuter = 245.;    //! outer radius of the TPC active area
```

Comment:
At some point we should check if these can go into a parameter class.

Reply:
Yes, this is also in the todo list.